src/screencast/pipewire_screencast.c

```diff
@@ -229,7 +229,27 @@ int xdpw_pwr_core_connect(struct xdpw_state *state) {
 	return 0;
 }
 
+void xdpw_pwr_context_destroy(struct xdpw_state *state) {
+	struct xdpw_screencast_context *ctx = &state->screencast;
+
```

Comment:
If we've gotten here, and there are still streams present, we either need to asynchronously defer execution of the remainder of this function until such a time when all streams have been terminated and cleaned up (likely due to the quit or err flags being set), or we need to manually iterate through all streams and destroy them synchronously.

The former approach would use a timer (some example of this in the commits related to the max fps PR). The latter approach would involve iterating through all streams and calling stream destroy for each (and setting their quit flags so that we don't reschedule them).

Reply:
Hmm, I can't really follow here. As far as I can see xdpw_pwr_context_destroy() is only called from xdpw_screencast_init() during the error path.

If that happens we either come from goto fail_pipewire, i.e. there shouldn't be any stream around anyway. Or we come from goto fail_screencopy, in which case xdpw_wlr_screencopy_finish() is called, and then xdpw_pwr_context_destroy().

In fact also xdpw_wlr_screencopy_finish() is only ever called from xdpw_screencast_init(), I don't see it referenced anywhere else.

Now you tell me to iterate over the screencast instances in xdpw_pwr_context_destroy() and clean them up. But that is precisely what is done in xdpw_wlr_screencopy_finish().

Maybe I'm missing the big picture here, but to me this just looks like moving code around, without actually changing the order of how things are done.

Reply:
I haven't had the time to actually test this or read anything beyond the diffs. I'll have a chance in the next couple weeks to give you a much more thorough review. I am making some assumptions based on @columbarius's comment, that the `-r` flag that does instance replacements hits this codepath when streams are active.

> The dead pipewire nodes are gone, which is nice, but when we replace xdpw with the -r flag pipewire clients can SEGFAULT, when streaming, which is not nice. We probably need a xdpw_screencast_finish, which goes the error path of xdpw_screencast_init and does quit all open screencast instances.

> This xdpw_screencast_finish should wait until all screencopy loops have finished (and destroyed the pipewire stream) and then destroy the context.

There is a race condition somewhere (that has always been an issue, just covered up by the fact that we've been leaving dead pipewire sessions around).

I can give you much more specific feedback later this week when I have time to read through the existing code again.

Thanks for your patience. We appreciate you working with us on this.

Reply:
Okay @tobiasjakobi, I've taken the time to understand what @columbarius was saying. Unrelated to your PR, we have an error condition [here](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/src/core/main.c#L263) that can occur with active streams hanging around. We would want to call this context_destroy function there as well.

This is more complicated than I first thought, because we are also outside of our event loop here, so using our timer setup isn't plausible. I think the right thing to do may just be to check if there are any lingering stream objects, and skip the destroy calls and return an error code if there are any. This would allow us to worry about how to use this function in a place where active streams *might* be present.

@columbarius I think our error label in `core/main.c` needs to be a second "simplified" event loop that can continue processing streams to allow us a chance to clean things up, but I don't want to go down that path on this PR. Do you think making `context_destroy` return int instead of void and checking for active streams first would be sufficient to accept this PR? I'd then propose we raise an issue about this and go back in a second PR and implement that improved exit condition.

Reply:
I think we should simplify this commit to just adding context_destroy and handle errors occurring in context_create.

Since with #137 I will restructure most of the screencast_instance stuff to rely more on pipewire, which has well defined signals for buffer creation and destruction. With that we don't have to wait until some loop reached an arbitrary point at which the cleanup is finished. I don't know, how much i have to touch, so i wouldn't spend to much time with that tbh.

Reply:
@columbarius In that case, I'm happy to close this review here and merge. Agree?